### PR TITLE
Refactor OnScreenDisplay to allow displaying on custom messages

### PR DESCRIPTION
**Pull request by Game4all**
This PR refactors the OnScreenDisplay overlay and adds a method allowing the display of custom on screen messages which are deriving from class `Toast` (one use case for this is #5268).

It moves all the drawables & the logic used to display tracked settings changes to it's own class (`TrackedSettingToast`)

### CODE REVIEW

**Comment by peppy** [comment]
I've tidied this up a bit, but it still requires larger changes. The text display should always use the same layout, so this portion can be moved to the base `Toast` class.

ie. for on-screen music controls, the display would be

```
____________________
|   music playback  |
|        pause      |
|         F5        |
--------------------
```

Reasoning:
- Icons you added feel clickable
- We still want to give the user context of the system they are interacting with
- We want to show the user which hotkey is related to the action they took
- Design consistency

So the base Toast class should likely take in its constructor the various strings required, and the only part moved to TrackedSettings would be the light logic.

Let me know if you'd prefer I complete these changes, else I will leave it to you.

**Comment by Game4all** [comment]
I should be able to do these changes on my own 😁

File: osu.Game/Overlays/OSD/TrackedSettingToast.cs [hunk 1]
```diff
@@ -90,9 +62,6 @@ public TrackedSettingToast(SettingDescription description)
                     break;
             }
 
-            textLine2.Origin = optionCount > 0 ? Anchor.BottomCentre : Anchor.Centre;
```
Comment (peppy) [inline osu.Game/Overlays/OSD/TrackedSettingToast.cs]: what happened to this?

Comment (Game4all) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: I deleted this intentionnally as it visually doesn't change anything now.

Comment (peppy) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: Hm, it should though. The vertical alignment would hopefully not change as a result of this PR>

Comment (Game4all) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: I did some verifications via visual tests to ensure it doesn't change anything before deleting it.

Comment (peppy) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: I'm not sure how that could work when this is specifically changing the alignment of the text line when there are options present...

Comment (peppy) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: here's how the vertical offset is currently applied for reference:

image
image

Comment (Game4all) [reply osu.Game/Overlays/OSD/TrackedSettingToast.cs]: looks like it was an attention error of mine when I moved the tracked setting's code to its own class, as it actually influences the display ,my bad

**Comment by peppy** [comment]
On a visual check, the option lights are moving incorrectly relative to the toast now

Previously (lights move down from centre):



On this branch (lights move up from centre):

**Comment by Game4all** [comment]
Keybing Sprite text at the bottom

File: osu.Game/Overlays/OnScreenDisplay.cs [hunk 2]
```diff
@@ -162,56 +91,23 @@ public void StopTracking(object source, ITrackableConfigManager configManager)
...
             existing.Unload();
-            existing.SettingChanged -= display;
+            existing.SettingChanged -= displayTrackedSettingChange;
 
             trackedConfigManagers.Remove((source, configManager));
         }
 
-        private void display(SettingDescription description)
+        /// <summary>
+        /// Displays the given <see cref="Toast"/> as parameter on the OSD
```
Comment (peppy) [inline osu.Game/Overlays/OnScreenDisplay.cs]:
```suggestion
        /// Displays the provided <see cref="Toast"/> temporarily.
```

Comment (peppy) [reply osu.Game/Overlays/OnScreenDisplay.cs]: why not just make `DisplayTemporarily` public and use that? this is the only call site of that function.

File: osu.Game/Overlays/OSD/Toast.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+namespace osu.Game.Overlays.OSD
+{
+    public abstract class Toast : Container
+    {
+        private readonly Container content;
+        protected override Container<Drawable> Content => content;
+
+        protected readonly OsuSpriteText ValueText;
+
+        protected Toast(string description, string value, string keybinding)
```
Comment (peppy) [inline osu.Game/Overlays/OSD/Toast.cs]:
```suggestion
        protected Toast(string description, string value, string shortcut)
```

it's called shortcut everywhere else

File: osu.Game/Overlays/OSD/Toast.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            Anchor = Anchor.Centre;
+            Origin = Anchor.Centre;
+            Width = 240;
```
Comment (peppy) [inline osu.Game/Overlays/OSD/Toast.cs]: broke width on long text

image

please write tests for edge cases like this.

File: osu.Game/Overlays/OSD/Toast.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+using osuTK.Graphics;
+
+namespace osu.Game.Overlays.OSD
+{
+    public abstract class Toast : Container
+    {
+        private readonly Container content;
+        protected override Container<Drawable> Content => content;
+
+        protected readonly OsuSpriteText ValueText;
```
Comment (peppy) [inline osu.Game/Overlays/OSD/Toast.cs]: why not `private`?

File: osu.Game/Overlays/OnScreenDisplay.cs [hunk 6]
```diff
@@ -8,34 +8,21 @@
...
-using osu.Framework.Graphics.Effects;
 using osu.Framework.Graphics.Transforms;
 using osu.Framework.Threading;
 using osu.Game.Configuration;
-using osu.Game.Graphics.Sprites;
+using osu.Game.Overlays.OSD;
 
 namespace osu.Game.Overlays
 {
     public class OnScreenDisplay : Container
```
Comment (peppy) [inline osu.Game/Overlays/OnScreenDisplay.cs]: needs xmldoc since you are overloading functionality.

"An on-screen display which automatically tracks and displays toast notifications for <see...TrackedSettings>. Can also display custom content via <see..Display>"

or something like this.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/OnScreenDisplay.cs (hunk 2 above)
Comment (Game4all) [reply osu.Game/Overlays/OnScreenDisplay.cs]: DisplayTemporarily contains mostly transformation code and is actually overriden in tests to show the toast without the fading out

image

File: osu.Game/Overlays/OSD/Toast.cs (hunk 5 above)
Comment (Game4all) [reply osu.Game/Overlays/OSD/Toast.cs]: This member is accessed in `TrackedSettingToast` to set the Anchor of the value sprite text in case of a setting with multiples options

image

File: osu.Game/Overlays/OSD/Toast.cs (hunk 4 above)
Comment (Game4all) [reply osu.Game/Overlays/OSD/Toast.cs]: I guess correct behaviour for this is having the toast autosizing on X axis

**Review by peppy — APPROVED** [review]
